### ISSUE TITLE
EcalHaloDataProducer causing circular dependencies in RECO path

### ISSUE THREAD

**Issue by Dr15Jones** [cms-sw/cmssw#15781]
The improved module/Path circular dependency checker found the following problem in workflow 8.0

```
photons consumes photonCore, photonCore after conversions [path reconstruction_step], conversions consumes ckfOutInTracksFromConversions, ckfOutInTracksFromConversions consumes conversionTrackCandidates, conversionTrackCandidates after BeamHaloSummary [path reconstruction_step], BeamHaloSummary consumes EcalHaloData, EcalHaloData consumes photons
```

The relevant part of the  `reconstruction_step` path looks like

```
EcalHaloData+HcalHaloData+GlobalHaloData+BeamHaloSummary+conversionTrackCandidates+ckfOutInTracksFromConversions+ckfInOutTracksFromConversions+conversions+photonCore+photons
```

Where `EcalHaloData` is the `EcalHaloDataProducer` and it does ask for the data from `photons`

```
>>> print process.EcalHaloData.dumpPython()
cms.EDProducer("EcalHaloDataProducer",
    AngleCutParam = cms.double(0.51),
    EBRecHitEnergyThresholdParam = cms.double(0.3),
    EBRecHitLabel = cms.InputTag("ecalRecHit","EcalRecHitsEB"),
    EERecHitEnergyThresholdParam = cms.double(0.3),
    EERecHitLabel = cms.InputTag("ecalRecHit","EcalRecHitsEE"),
    ESRecHitEnergyThresholdParam = cms.double(0.3),
    ESRecHitLabel = cms.InputTag("ecalPreshowerRecHit","EcalRecHitsES"),
    HBHERecHitLabel = cms.InputTag("hbhereco"),
    NHitsEcalThresholdParam = cms.int32(4),
    PhotonLabel = cms.InputTag("photons"),
    RoundnessCutParam = cms.double(0.41),
    SumEcalEnergyThresholdParam = cms.double(10.0),
    SuperClusterLabel = cms.InputTag("correctedHybridSuperClusters")
)
```

`EcalHaloDataProducer` uses `EcalHaloAlgo` to do the work and that class ignores the Photon collection if the handle is invalid
https://github.com/cms-sw/cmssw/blob/f1191950193ae671897f0484ff4cec28281a9da1/RecoMET/METAlgorithms/src/EcalHaloAlgo.cc#L156

If this path were fully converted to unscheduled than `EcalHaloDataProducer` would return a different answer than it does now (since it would get the data from `photons`). That is highly undesirable and the new checking algorithm will no longer allow that.

I see two possible fixes
1.  move `EcalHaloData` (and anything that depends on it) to be after `photons` in the path
2.  explicitly tell `EcalHaloData` to not get the photons, i.e.

```
EcalHaloData.PhotonLabel = cms.InputTag("")
```

**Comment by Dr15Jones** [comment]
This problem must be resolved before I can commit the next threaded framework change since that change requires that no circular dependencies exist in the schedule.

**Comment by Dr15Jones** [comment]
assign reconstruction

**Comment by slava77** [comment]
@lathomas @mariadalfonso  @zdemirag 
please investigate

@Dr15Jones I suppose, this is essential at this point for 81X and later, right?

**Comment by slava77** [comment]
@Dr15Jones if we just change 8.0 to run this step unscheduled, will you be able to continue with your threaded framework changes?

**Comment by Dr15Jones** [comment]
Unscheduled would fix this (assuming there isn't a true underlying cycle from data dependencies lurking in that configuration).

**Comment by lathomas** [comment]
@slava77 @Dr15Jones  It's fine to set the photon collection to an empty string. The collection is not used in the current halo filter.

**Comment by slava77** [comment]
On 9/9/16 5:21 AM, lathomas wrote:
> @slava77 <https://github.com/slava77> @Dr15Jones
> <https://github.com/Dr15Jones> It's fine to set the photon collection to
> an empty string. The collection is not used in the current halo filter.

Would you like to make a pull request for this?
Please let us know.

Thank you.

**Comment by slava77** [comment]
+1

fixed in #15799
